gobblin-modules/gobblin-compliance/src/main/java/gobblin/compliance/retention/ComplianceRetentionJob.java

```diff
@@ -59,28 +66,52 @@
  * A compliance job for compliance retention requirements.
  */
 @Slf4j
+@SuppressWarnings
 public class ComplianceRetentionJob extends ComplianceJob {
-  public static final List<HiveDataset> datasetList = new ArrayList<>();
+  public static final List<String> tableNamesList = new ArrayList<>();
+  public static HiveMetastoreClientPool pool;
+
   public ComplianceRetentionJob(Properties properties) {
     super(properties);
-    initDatasetFinder(properties);
     try {
-      Iterator<HiveDataset> datasetsIterator =
-          new HiveDatasetFinder(FileSystem.newInstance(new Configuration()), properties).getDatasetsIterator();
-      while (datasetsIterator.hasNext()) {
-        this.datasetList.add(datasetsIterator.next());
-      }
+      this.pool = HiveMetastoreClientPool
+          .get(properties, Optional.fromNullable(properties.getProperty(HiveDatasetFinder.HIVE_METASTORE_URI_KEY)));
+      initDatasetFinder(properties);
       ProxyUtils.cancelTokens(new State(properties));
     } catch (InterruptedException | TException | IOException e) {
       Throwables.propagate(e);
     }
   }
 
-  public void initDatasetFinder(Properties properties) {
+  public void initDatasetFinder(Properties properties)
+      throws IOException {
     Preconditions.checkArgument(properties.containsKey(GOBBLIN_COMPLIANCE_DATASET_FINDER_CLASS),
         "Missing required propety " + GOBBLIN_COMPLIANCE_DATASET_FINDER_CLASS);
     String finderClass = properties.getProperty(GOBBLIN_COMPLIANCE_DATASET_FINDER_CLASS);
     this.finder = GobblinConstructorUtils.invokeConstructor(DatasetsFinder.class, finderClass, new State(properties));
+
+    Iterator<HiveDataset> datasetsIterator =
+        new HiveDatasetFinder(FileSystem.newInstance(new Configuration()), properties).getDatasetsIterator();
+
+    while (datasetsIterator.hasNext()) {
+      // Drop partitions from empty tables if property is set, otherwise skip the table
+      HiveDataset hiveDataset = datasetsIterator.next();
+      List<Partition> partitionsFromDataset = hiveDataset.getPartitionsFromDataset();
```

Comment:
The general convention is not to delete during iteration. Could you please keep a list of the tables you'd like to drop and do so separately after the init method, perhaps?